test/13_arraycommon.jl

```diff
@@ -184,6 +184,43 @@ end
             @test_throws BoundsError copy!(x, 1, y, 4, 2)
             @test x ≅ a
         end
+
+        @testset "level preservation with copy!" begin
+            v = ["a", "b", "c"]
+            src = CategoricalVector(v)
+            dest = CategoricalVector{String}(3)
+            @test levels(copy!(dest, src)) == v
+
+            src = levels!(CategoricalVector(v), reverse(v))
+            dest = CategoricalVector{String}(3)
+            @test levels(copy!(dest, src)) == reverse(v)
+
+            dest = CategoricalVector{Union{String, Null}}(3)
+            @test levels(copy!(dest, src)) == reverse(v)
+
+            dest = CategoricalVector{Union{String, Null}}(3)
+            dest[1] = src[1]
+            @test levels(copy!(dest, 2, src[2:end])) == reverse(v)
+
+            src = levels!(CategoricalVector{Union{String, Null}}(v), reverse(v))
+            dest = CategoricalVector{Union{String, Null}}(3)
+            @test levels(copy!(dest, src)) == reverse(v)
+
+            src = levels!(CategoricalVector{Union{String, Null}}(v), reverse(v))
+            dest = CategoricalVector{String}(3)
+            @test levels(copy!(dest, src)) == reverse(v)
+
+            # incorrectly leaves #undef in 1st entry of dest
+            src = levels!(CategoricalVector{Union{String, Null}}(v), reverse(v))
+            src[1] = null
+            dest = CategoricalVector{String}(3)
+            @test levels(copy!(dest, src)) == reverse(v)
+
+            # does not dispatch correctly
+            dest = CategoricalVector{Union{String, Null}}(3)
+            vsrc = view(src, 1:length(src))
+            @test levels(copy!(dest, vsrc)) == reverse(v)
```

Comment:
I've been stuck on this for a few days now, not sure what's going on. I can't get this to dispatch to the full, 5-argument `copy!` function. It goes through the `copy!(dest::CategoricalArray, src::CA_OR_CA_VIEW)` function fine and `vsrc` in this test checks out as being `<: CA_OR_CA_VIEW` but `copy!(dest, 1, src, 1, length(src))` within the 2-argument helper function (line 442) is calling some other `copy!`. Any ideas? We could make second 5-argument `copy!` function just for `src::SubArray{CategoricalArray}` but I'm sure we can come up with something more elegant than that

Reply:
I guess it might be calling a method for `Base.LinearSlow` or `CartesianRange`, like `copy!( ::LinearFast, ::CategoricalArray, ::LinearSlow, ::SubArray{CategoricalArray})` or `copy!(::AbstractArray{T1,N}, ::CartesianRange{CartesianIndex{N}}, ::AbstractArray{T2,N}, ::CartesianRange{CartesianIndex{N}}) where {T1, T2, N}`.